packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts

```diff
@@ -64,18 +64,45 @@ export function createPgDatabaseClient(
  * @param dbConfig - The database config
  * @param overrides - Additional options to merge with the config
  */
-export function buildPgDatabaseConfig(
+export async function buildPgDatabaseConfig(
   dbConfig: Config,
   overrides?: Knex.Config,
 ) {
-  return mergeDatabaseConfig(
+  const config = mergeDatabaseConfig(
     dbConfig.get(),
     {
       connection: getPgConnectionConfig(dbConfig, !!overrides),
       useNullAsDefault: true,
     },
     overrides,
   );
+
+  if (config.client === 'pg+google-cloudsql') {
+    const {
+      Connector: CloudSqlConnector,
+      IpAddressTypes,
+      AuthTypes,
+    } = await import('@google-cloud/cloud-sql-connector');
+    // override the config to be pg for backwards compat with other code
+    config.client = 'pg';
+
+    const connector = new CloudSqlConnector();
+    const clientOpts = await connector.getOptions({
+      instanceConnectionName: dbConfig.getString('instanceConnectionName'),
```

Comment:
note that this and the above new client string should have a corresponding `config.d.ts` [update](https://github.com/backstage/backstage/blob/03351abbddbc5bbee85afcdcbc99517af199c7ee/packages/backend-defaults/config.d.ts#L380). Actually, the instance connection name (and possibly the ip and auth types etc) should probably go in the connection object in the future.

```yaml
client: 'pg+google-cloudsql'
connection: 'the-cloudsql-connection-identifier-thing'
```

as a shorthand and then

```yaml
client: 'pg+google-cloudsql'
connection:
  instance: 'the-cloudsql-connection-identifier-thing'
  ip: public
  ...
```

as the long form when we add support for more parameters